src/pip/_internal/index.py

```diff
@@ -226,6 +228,136 @@ def _get_html_page(link, session=None):
         return HTMLPage(resp.content, resp.url, resp.headers)
 
 
+class _LinkParseError(ValueError):
+    """Base class for all exceptions raised by `_parse_package_version()`.
+
+    The callee is expected to catch this, and inspect attributes `link` or
+    `search`, or convert the exception to a string for logging.
+    """
+    _template = "Missing project version for {search.supplied}"
+
+    def __init__(self, link, search):
+        super(_LinkParseError, self).__init__(link, search)
+        self.link = link
+        self.search = search
+
+    def __str__(self):
+        return self._template.format(link=self.link, search=self.search)
+
+
+class _LinkNotAFile(_LinkParseError):
+    _template = "not a file"
+
+
+class _LinkFormatUnsupported(_LinkParseError):
+    _template = "unsupported archive format: {link.ext}"
+
+
+class _MacOSXZip(_LinkParseError):
+    _template = "macosx10 one"
+
+
+class _SourceLinkNotPermitted(_LinkParseError):
+    _template = "No sources permitted for {search.supplied}"
+
+
+class _BinaryLinkNotPermitted(_LinkParseError):
+    _template = "No binaries permitted for {search.supplied}"
+
+
+class _InvalidWheelFilename(_LinkParseError):
+    _template = "invalid wheel filename"
+
+
+class _IncorrectProjectName(_LinkParseError):
+    _template = "wrong project name (not {search.supplied})"
+
+
+class _PythonNotSupported(_LinkParseError):
```

Comment:
I wouldn't say this is a parse error.
I don't really see the point of having 10 different internal exception types: maybe a single `SkippedLink(link, reason)` exception would be sufficient ?